### PR TITLE
Use correct type for Tk addresses.

### LINKED ISSUE #7633: All attempts to plot fail with "OverflowError: Python int too large to convert to C long"
To help us understand and resolve your issue please check that you have provided
the information below.

- [x] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...)

Platform: Debian Testing, using the python 2.7 (2.7.11-2 when I first started seeing this, just upgraded to 2.7.13rc1 but the problem still exists). I have this problem with the 1.5.3 on debian testing, but I also tried a virtualenv with pip install --pre matplotlib, which pulled in 2.0.0rc1, and I see the same problem.

Debian stable, with matplotlib 1.4.2, works fine, as did Debian testing until recently.

- [x] How did you install Matplotlib and Python (pip, anaconda, from source ...)

See above: Debian package and virtualenv.

- [x] If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
      that demonstrates the issue i.e a small piece of code which reproduces the issue
      and can be run with out any other (or as few as possible) external dependencies.

Minimal test program:

import pylab

distances = [0, 0.0011970407617050934, 0.0050796062254631295, 0.007303941153223762, 0.007303941153223762, 0.008232870398475445, 0.008232870398475445, 0.012806461740481582, 0.016651523629144134, 0.021047330616832543]
elevations = [6404.2, 6410.76, 6400.92, 6410.76, 6423.88, 6423.88, 6453.41, 6410.76, 6427.17, 6420.6]
pylab.plot(distances, elevations)
pylab.show()

(Please let me know if you want a non-pylab example. I see it with straight matplotlib too.)

- [x] If this is an image generation bug attach a screenshot demonstrating the issue.
- [x] If this is a regression (Used to work in an earlier version of Matplotlib), please 
      note where it used to work.

It worked as recently as a month ago on Debian testing, but sorry, I don't know what matplotlib version that was since I didn't know it was going to stop working.

### PR DESCRIPTION
**Pull request by QuLogic**
CPython has used a long for this address since before 2010 (it is difficult to read the merge from SVN back then), and has been a long-from-void* since 2013 (3.3.3 and 2.7.6).

Fixes #7633.

Also, remove the buggy and unused `_pyobj_addr` which really is not very portable.

I don't have a 32-bit system to actually verify this, but based on the upstream bug report maybe 64-bit Windows could fall prey to this as well. @akkana, please try out this branch if possible.

### CODE REVIEW

**Comment by tacaswell** [comment]
Raising at the c level is typically a critical bug (as it is extra hard for users to debug/patch around) which is why i moved this to 2.0 mile stone

**Comment by tacaswell** [comment]
attn @matthew-brett

**Comment by akkana** [comment]
It works! I built (and got version 2.0.0rc1+79.ga91559b -- I hope that means I applied the pull request right :-) and now in the venv my matplotlib programs are displaying graphs.

Thanks!

**Comment by tacaswell** [comment]
Also related https://bugs.python.org/issue18909 (@qulogic linked to the resulting patch).

My understanding of what is going wrong here (sorry for transcribing some way to picky details):

 - we use `tkapp.interpaddr` to get the memory address of the underlying tk app (so we can get a pointer to it to pass back into tcl to register our custom method onto the interpreter to blit the plot into the GUI)
 - under the hood (as of 2013 via https://github.com/python/cpython/commit/bef29fb3fa3118c63dfd8ba4baa13f2fed055fba) this is implemented as

```c
static PyObject *
Tkapp_InterpAddr(PyObject *self, PyObject *args)
{

    if (!PyArg_ParseTuple(args, ":interpaddr"))
        return NULL;

    return PyLong_FromVoidPtr(Tkapp_Interp(self));
}
```

Prior to this (back to when it was added in 1998) this was implemented as (in https://github.com/python/cpython/commit/83e9f84893946b7409d1f8b32e5a14e82ab000cf )

```c
static PyObject *
Tkapp_InterpAddr(PyObject *self, PyObject *args)
{

    if (!PyArg_ParseTuple(args, ":interpaddr"))
        return NULL;

    return PyInt_FromLong((long)Tkapp_Interp(self));
}
```

The change is due to a `long` on win64 being 32bits, but pointers being 64bit (and @cgohlke preemptively saving us from segfaults :+1: )

 - previously we were casting this to an `Py_ssize_t` in the tuple arg parsing step (as that is what the commit message said to do :wink: ) and this works out so long as the integer size and the pointer size line up.  In #7633 for some reason the pointer coming back is out of range the 32bit range, despite being on a 32bit system.  @sandrotosi can you shed any light on what is going on here?  Is this a debian cross-compile issue?

 - given the current implementation of `Tkapp_InterpAddr`, switching to `PyLong_AsVoidPtr` is clearly the correct behavior.  These come in to python in 1.5.2 so all of our versions are fully supported

 - `PyLong_AsVoidPtr` will properly up-cast a `PyInt` to a `PyLong` so versions of python prior to the above change will also behave correctly.

So, in conclusion, I am convinced that this works in principle and the OP reports that it works in practice so merging.

**Comment by QuLogic** [comment]
Based on the issue, the limit may have been assumed to be 31 bits, so if Python/kernel's got some address fuzzing going on (ASLR, maybe), it could produce a 32-bit address that wouldn't fit. I'm not entirely sure about that.
